src/ruby/pb/test/xds_client.rb

```diff
@@ -0,0 +1,193 @@
+      index = $watchers.length
+      # Add a watcher for this range of id's
+      $watchers << {
+        "start_id" => start_id,
+        "end_id" => end_id,
+        "rpcs_by_peer" => Hash.new(0),
+        "rpcs_needed" => req['num_rpcs'],
+        "no_remote_peer" => 0
+      }
+    end
+    # wait until all RPCs are done, or we are past timeout
+    while ($watchers[index]['rpcs_needed'] > 0) &&
+          (Time.now.to_f < finish_time)
+      sleep(0.1)
+    end
+    LoadBalancerStatsResponse.new(
+      rpcs_by_peer: $watchers[index]['rpcs_by_peer'],
+      num_failures: $watchers[index]['no_remote_peer'] +
+      $watchers[index]['rpcs_needed']
+    );
+  end
+end
+
+# send 1 rpc every 1/qps second
+def run_test_loop(stub, duration_per_query)
+  include Grpc::Testing
+  req = SimpleRequest.new()
+
+  while true
+    start = Time.now.to_f
+    begin
+      resp = stub.unary_call(req)
+      remote_peer = resp.hostname
+    rescue GRPC::Unavailable => e
+      remote_peer = ""
+    end
+    finish = Time.now.to_f
+    delay = duration_per_query - (finish - start)
+    if delay > 0
+      sleep(delay)
+    else
+      GRPC.logger.info("ruby xds: warning, rpc takes too long to finish. " \
+                       "If you consistently see this, the qps is too high.")
+    end
+    $current_request_id += 1
+    $watchers_mutex.synchronize do
+      $watchers.each do |watcher|
+        if (watcher['start_id'] <= $current_request_id) &&
+           ($current_request_id < watcher['end_id'])
+          watcher['rpcs_needed'] -= 1
+          if remote_peer.strip.empty?
+            watcher['no_remote_peer'] += 1
+          else
+            watcher['rpcs_by_peer'][remote_peer] += 1
+          end
+        end
+      end
+    end
+  end
+end
+
+# Args is used to hold the command line info.
+Args = Struct.new(:server, :stats_port, :qps)
+
+# validates the command line options, returning them as a Hash.
+def parse_args
+  args = Args.new
+  OptionParser.new do |opts|
+    opts.on('--server SERVER_HOST', 'server hostname') do |v|
+      GRPC.logger.info("ruby xds: server address is #{v}")
+      args['server'] = v
+    end
+    opts.on('--stats_port STATS_PORT', 'stats port') do |v|
+      GRPC.logger.info("ruby xds: stats port is #{v}")
+      args['stats_port'] = v
+    end
+    opts.on('--qps QPS', 'qps') do |v|
+      GRPC.logger.info("ruby xds: qps is #{v}")
+      args['qps'] = v
+    end
+  end.parse!
+  args
+end
+
+def main
+  opts = parse_args
+
+  # This server hosts the LoadBalancerStatsService
+  host = "0.0.0.0:#{opts['stats_port']}"
+  s = GRPC::RpcServer.new
+  s.add_http2_port(host, :this_port_is_insecure)
+  s.handle(TestTarget)
+  server_thread = Thread.new {
+    s.run_till_terminated
+  }
+
+  # The client just sends unary rpcs continously in a regular interval
+  stub = create_stub(opts)
+  duration_per_query = (1.0 / opts['qps'].to_f)
+  client_thread = Thread.new {
+    run_test_loop(stub, duration_per_query)
+  }
+
```

Comment:
let's make sure to join the `client_threads`

I think we could do that by adding a `shutdown` global variable, setting it true here, and checking it at the top of each iteration of the while loop, returning from the loop if shutdown is true.

No need to guard `shutdown` with a mutex since it's a single variable thereby protected by the GIL.

Reply:
The Python test runner is terminating this Ruby client by `subprocess.terminate()`, so a `SIGTERM` signal is being sent to this program. So I modified the server to start with `s.run_till_terminated_or_interrupted(['TERM'])`. And it successfully causes `server_thread.join` to finally return when the Python test runner attempts to terminate this program. After that point, I am able to set the `$shutdown` global variable to cause each of the `client_thread` to terminate and hence they can all `client_thread.join` too. And I verified that we reached the last line of the `main` function properly to shut down.

Reply:
thanks, this looks good!